src/transactions/TransactionFrame.cpp

```diff
@@ -57,6 +58,112 @@ namespace
+// to be consumed by downstream systems. In particular, restoration is
+// (mostly) logically a new entry creation from the perspective of ltx and
+// stellar-core as a whole, but this change type is reclassified to
+// LEDGER_ENTRY_RESTORED for easier consumption downstream.
+LedgerEntryChanges
+processOpLedgerEntryChanges(std::shared_ptr<OperationFrame const> op,
+                            AbstractLedgerTxn& ltx)
+{
+    if (op->getOperation().body.type() != RESTORE_FOOTPRINT)
+    {
+        return ltx.getChanges();
+    }
+
+    auto const& hotArchiveRestores = ltx.getRestoredHotArchiveKeys();
+    auto const& liveRestores = ltx.getRestoredLiveBucketListKeys();
+
+    LedgerEntryChanges changes = ltx.getChanges();
+
+    // Depending on whether the restored entry is still in the live
+    // BucketList (has not yet been evicted), or has been evicted and is in
+    // the hot archive, meta will be handled differently as follows:
+    //
+    // Entry restore from Hot Archive:
+    // Meta before changes:
+    //     Data/Code: LEDGER_ENTRY_CREATED
+    //     TTL: LEDGER_ENTRY_CREATED
+    // Meta after changes:
+    //     Data/Code: LEDGER_ENTRY_RESTORED
+    //     TTL: LEDGER_ENTRY_RESTORED
+    //
+    // Entry restore from Live BucketList:
+    // Meta before changes:
+    //     Data/Code: no meta
+    //     TTL: LEDGER_ENTRY_STATE(oldValue), LEDGER_ENTRY_UPDATED(newValue)
+    // Meta after changes:
+    //     Data/Code: LEDGER_ENTRY_RESTORED
+    //     TTL: LEDGER_ENTRY_STATE(oldValue), LEDGER_ENTRY_RESTORED(newValue)
+    //
+    // First, iterate through existing meta and change everything we need to
+    // update.
+    for (auto& change : changes)
+    {
+        // For entry creation meta, we only need to check for Hot Archive
+        // restores
+        if (change.type() == LEDGER_ENTRY_CREATED)
+        {
+            auto le = change.created();
+            if (hotArchiveRestores.find(LedgerEntryKey(le)) !=
+                hotArchiveRestores.end())
+            {
+                releaseAssertOrThrow(isPersistentEntry(le.data) ||
+                                     le.data.type() == TTL);
+                change.type(LEDGER_ENTRY_RESTORED);
+                change.restored() = le;
+            }
+        }
+        // Update meta only applies to TTL meta
+        else if (change.type() == LEDGER_ENTRY_UPDATED)
+        {
+            if (change.updated().data.type() == TTL)
+            {
+                auto ttlLe = change.updated();
+                if (liveRestores.find(LedgerEntryKey(ttlLe)) !=
+                    liveRestores.end())
+                {
+                    // Update the TTL change from LEDGER_ENTRY_UPDATED to
+                    // LEDGER_ENTRY_RESTORED.
+                    change.type(LEDGER_ENTRY_RESTORED);
+                    change.restored() = ttlLe;
+                }
+            }
+        }
+    }
+
+    // Now we need to insert all the LEDGER_ENTRY_RESTORED changes for the
```

Comment:
I don't quite understand this, aren't we adding duplicate changes here?

Reply:
I've updated the comment explain this a bit more. When an entry still exists on the BucketList, all the restoration needs to do is change the TTL value, so we never actually change the code/data entry. This means that the ltx only has meta events for the TTL change. In this case, platform requested that we also emit the entry being restored in meta even though it hasn't changed, so we add it manually here. Otherwise, they'd have to do DB lookups during meta ingest to determine what TTL restore events map to which entries.

Reply:
Thanks, I've missed that we have only TTL meta above.